sky/data/sky_csync.py

```diff
@@ -0,0 +1,545 @@
+
+
+def _terminate(paths: List[str], all: bool = False) -> int:  # pylint: disable=redefined-builtin
+    """Terminates the CSYNC daemon running.
+
+    Before terminating the running CSYNC daemon, it checks if the sync process
+    spawned by the daemon is running. If it is, we wait until the sync gets
+    completed, and then proceed to terminate the daemon.
+    """
+    csync_pid_set = set()
+    if all:
+        for csync_pid in _get_all_running_csync_pid():
+            csync_pid_set.add(csync_pid)
+    # when terminating specified CSYNC processes given the paths where
+    # the processes are running on.
+    else:
+        for path in paths:
+            full_path = os.path.abspath(os.path.expanduser(path))
+            csync_pid = _get_csync_pid_from_mountpoint_path(full_path)
+            if csync_pid is not None:
+                csync_pid_set.add(csync_pid)
+
+    pid_set = set()
+    for csync_pid in csync_pid_set:
+        storage_mount_pid, redirect_mount_pid = (
+            _get_storage_and_redirect_mount_pid(csync_pid))
+        pid_set.add((csync_pid, storage_mount_pid, redirect_mount_pid))
+
+    failed_to_terminate: bool = False
+    failed_to_terminate_list: List[Tuple[int, str, str]] = []
+    unmount_cmd = 'sudo fusermount -uz {unmount_path}'
+    for csync_pid, storage_mount_pid, redirect_mount_pid in pid_set:
+        # 1.unmount cloud storage from read path
+        csync_read_path = os.path.abspath(
+            os.path.expanduser(_CSYNC_READ_PATH.format(pid=csync_pid)))
+        storage_fuse_unmount_cmd = unmount_cmd.format(
+            unmount_path=csync_read_path)
+        _, rc, stderr = _handle_fuse_process(storage_fuse_unmount_cmd)
+        if rc != 0:
+            failed_to_terminate = True
+            failed_to_terminate_list.append(
+                (csync_pid, stderr, 'Storage Mount FUSE'))
+            continue
+
+        # 2.unmount redirection FUSE
+        mountpoint_path = _get_mountpoint_path_from_csync_pid(csync_pid)
+        redirection_fuse_unmount_cmd = unmount_cmd.format(
+            unmount_path=mountpoint_path)
+        _, rc, stderr = _handle_fuse_process(redirection_fuse_unmount_cmd)
+        if rc != 0:
+            failed_to_terminate = True
+            failed_to_terminate_list.append(
+                (csync_pid, stderr, 'Redirection FUSE'))
+            continue
+        # If there's any process obtaining the file descriptor of a file in
+        # mountpoint, i.e. writing a checkpoint, the mountpoint will remain
+        # mounted until the file descriptor is released after running
+        # fusermount -uz MOUNT_PATH. We can confirm if the mountpoint is
+        # completely unmounted by checking the termination
+        # of the mount process.
+        while True:
+            try:
+                process = psutil.Process(redirect_mount_pid)
+                # Check if process is still running
+                if (process.status() == psutil.STATUS_ZOMBIE or
```

Comment:
do we need to kill the zombie process

Reply:
Hm should we leave them?.. I can't think of a good reason to leave them. What do you think?

Reply:
Lets kill them if it does not require a lot of code changes?

Reply:
Oh I misunderstood what you meant. Yeah let's definitely kill the zombie process.

Reply:
Oh makes sense. nvm